### DISCUSSION TITLE
ServiceRequest search by `resolve()`

### DISCUSSION THREAD

**Discussion by whyfate** [FirelyTeam/firely-net-sdk#1772] [Q&A] [answered]
**Describe the bug**
```csharp
var sr = new ServiceRequest
{
    Subject = new ResourceReference
    {
        Reference="Patient/1",
    }
};
// values Count is 0
var values = sr.ToTypedElement().Select("ServiceRequest.subject.where(resolve() is Patient)");
```
But the similar cases
```csharp
Patient patient = new()
{
    Id = "1",
    Active = true,
    BirthDate = "2001-03-01"
};

Encounter encounter = new()
{
    Contained = new List<Resource>() { patient },
    Subject = new ResourceReference("#1")
};
// values Count is 1
var values = encounter.Select("Encounter.subject.where(resolve() is Patient)");
```

I use 3.3.0 version of  `firely-net-sdk`.

**Comment by marcovisserFurore** [comment]
Hi @whyfate ,

In your second example FhirPath can resolve the `Patient`, because this resource is contained in the `Encounter` resource. So FhirPath is able to find this resource. In your first example the `ServiceRequest` is reference to a non-contained resource. That means that the resource can be on the same server, but even on a complete other server. FhirPath cannot determine that.

To overcome that, you can tell FhirPath how to resolve resource, like this:

```c#
            var context = FhirEvaluationContext.CreateDefault();

            var client = new FhirClient("https//fhir.server.example");
            var resolver = new WebResolver((uri) => client);

            context.ElementResolver = (s) => resolver.ResolveByCanonicalUri(s).ToTypedElement();
            var values = sr.Select("ServiceRequest.subject.where(resolve() is Patient)", context);
```

In this case we tell FhirPath to request the server "https//fhir.server.example" to retrieve the resource. Assuming the resource (`ServiceRequest`) is on that server.

**Comment by whyfate** [comment]
It's so complicated.

**Comment by marcovisserFurore** [comment]
It is. But maybe you can explain your use case, and then there might be a better solution. For example, is your use case to retrieve ServiceRequest resources from a Fhir server with subject is Patient? Then you can write a search request to the server a let the server do the filtering.

**Comment by whyfate** [comment]
I use [spark](https://github.com/FirelyTeam/spark) as fhir server.
I read the docs(http://build.fhir.org/servicerequest.html#search), I want to search the `ServiceRequest` by `patient`.
![image](https://user-images.githubusercontent.com/13515020/122744644-c2ddd500-d2ba-11eb-956d-16ac466b975a.png)

`spark` is not implemented.
I tried to modify the code, so complicated.

**Comment by kennethmyhra** [comment]
I have never worked with resolvers in the firely-net-sdk library. For Spark would probably need to implement a custom resource resolver (IAsyncResourcesResolver/IResourceResolver) that can resolve resources from the database.

**Comment by whyfate** [comment]
@kennethmyhra It's will be slowly.Maybe I don't want to verify whether it really exists.

**Comment by kennethmyhra** [comment]
There exists no such resolver server-side in Spark today.

But I think Marco's example using the WebResolver client-side should be quite easy to implement without having to modify Spark's source code. Speed will of course depend on how many such resolve call you need to issue.

**Comment by whyfate** [comment]
Emmm...but isn't that what the server should provide?

**Comment by kennethmyhra** [comment]
I have not worked much with FHIRPath and resolve, but I don't think there is a good way in the specification to issue such a FHIRPath to the server from a client?

**Comment by whyfate** [comment, marked as answer]
@kennethmyhra @marcovisserFurore  Thank you very much for your answers.
I'll use `subject` to instead of `patient` to search `ServiceRequest`.
This is best way for me right now.

**Reply by marcovisserFurore** [reply]
I moved this issue to the discussion section, because it is not related to a bug in the SDK, but can be used for future references.
